docs/source/features/optimizations/sequence_packing.rst

```diff
@@ -138,7 +138,7 @@ Sequence Packing for NeVA
 Sequence packing in NeVA (Multimodal LLMs) differs slightly from the LLM SFT/PEFT approach. For details,
 please refer to the documentation below
 
-:doc:`../multimodal/mllm/sequence_packing`
+:doc:`../../multimodal/mllm/sequence_packing`
 
```

Comment:
copyedits for docs/source/features/optimizations/sequence_packing.rst

Line 3. add introductory text to this topic.

This section explains how to use the sequence packing training technique with Supervised Fine-Tuning (SFT) and Parameter-Efficient Fine-Tuning (PEFT).

Lines 7-8 revise sub-headline format.

Overview
########

Lines 10 -14. fix punctuation and revise.

When fine-tuning a large language model, whether using SFT or PEFT methods, GPU underutilization often occurs due to an inefficient data pipeline. This inefficiency arises because most fine-tuning datasets have a skewed distribution of sequence lengths, with many short sequences and a few long ones, following Zipf’s Law. Since transformer models require fixed-length inputs, shorter sequences must be padded with unused tokens, leading to two main inefficiencies:

Lines 20 - 22. fix punctuation and revise.

Sequence packing is a training technique where multiple training sequences (examples) are concatenated into one long sequence (pack). This method eliminates the need for padding, allowing more tokens to be processed in each micro batch. As a result, it maximizes both GPU compute and GPU memory utilization.

Line 30 revise.

With this approach, attention values between sequences are never calculated,


Lines  42 - 43. revise sub-headline format and fix capitalization.

Run SFT/PEFT with Packed Sequence
##################################

Line 44. add introductory text to this section.

This section explains how to prepare the dataset and adjust the training configuration.

Lines 45 - 46. revise sub-sub-headline format.

Prepare the Dataset
^^^^^^^^^^^^^^^^^^^

Line  51. fix capitalization.

download and prepare the Dolly dataset as an example.

Lines  57 - 59. suggested revision.

1. The online processing code in GPTSFTDataset is run. This includes tasks such as prompt template manipulation, sequence length truncation, and tokenization. The result is an array of tokenized sequences, represented by indices.

2. The tokenized sequences are grouped by length and a packing algorithm is run.

Lines 62 - 67. fix punctuation and formatting. The "-" character and the "*" character don't render correctly when used together. suggested revision.

Currently, two variants of ``first fit`` are supported:

* ``first_fit_decreasing`` sorts the sequences in decreasing order before applying the first-fit algorithm. It generates a more optimal packing, but it tends to keep all short sequences together, which may have an impact for convergence.

* ``first_fit_shuffle`` runs first-fit in a random order. Packing is less optimal but it keeps the dataset order random.

The recommendation is to run ``first_fit_shuffle`` and check the packed sequence lengths. If they are similar to the
target length (i.e. efficient packing), then use shuffle. Otherwise try ``first_fit_decreasing``.

Lines 82 -83.  Change "Note 1. to "1." and revise.

    1. If your model or dataset requires non-default configs for conventional SFT/PEFT training in NeMo, you will
    need to pass in the same configs to ``model.data.train_ds`` as you would for training with an unpacked dataset.

Lines 85 - 87. change "Note 2." to "2.", fix punctuation, and revise.

  2. ``model.data.train_ds.max_seq_length`` is the length to which each sequence is truncated before packing multiple sequences
    to the size of packed sequence (``pack_size``). ``max_seq_length`` should be set to the same value as unpacked data
    and can be determined by examining the distribution of sequence lengths in the dataset.

Lines 89 - 90. change "Note 3." to "3." .

  3. ``pack_sizes`` is a list of packed sequence lengths. In this example, there will be three output files, one for
    each pack size.

Lines 96 - 97.  revise sub-sub-headline format and title.

Adjust the Training Configuration
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Line 99. fix punctuation.

To train with packed sequences, you need to change four items in the SFT/PEFT config file.

Line 101. fix punctuation.

1. Turn on the packed_sequence flag:

Line 107. fix punctuation.

2. Use the new dataset file instead of the original jsonl file:

Line 133. fix punctuation.

Now, you are all set to fine-tune your model with a much improved throughput!

Lines 138 - 139. define acronymn and fix punctuation.

Sequence packing with Neural Variational Autoencoder (NeVA) for multimodal large language models differs from the LLM SFT/PEFT approach. For details, please refer to the documentation below.